### DISCUSSION TITLE
Building DDS-Router

### DISCUSSION THREAD

**Discussion by iekhanx** [eProsima/DDS-Router#188] [Q&A]
Hi there,
I am trying to build dds-router. I followed 
https://github.com/eProsima/DDS-Router
provided procedure but system generates following errors:
can you please advise how to over come from this error?
Thank you,
# colcon build
Starting >>> ddsrouter_cmake
Starting >>> fastcdr
Starting >>> foonathan_memory_vendor
Starting >>> ddsrouter_yaml_validator
Finished <<< fastcdr [0.35s]                                         
Finished <<< foonathan_memory_vendor [0.39s]                           
Finished <<< ddsrouter_cmake [0.47s]
Starting >>> fastrtps
Starting >>> ddsrouter_docs
--- stderr: ddsrouter_yaml_validator                                    
/usr/lib/python3/dist-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
  warnings.warn(
---
Finished <<< ddsrouter_yaml_validator [0.94s]
--- stderr: ddsrouter_docs                                    
WARNING: sphinx_rtd_theme (< 0.3.0) found. It will not be available since Sphinx-6.0

Theme error:
no theme named 'sphinx_rtd_theme' found (missing theme.conf?)
gmake[2]: *** [CMakeFiles/Sphinx.dir/build.make:71: CMakeFiles/Sphinx] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:110: CMakeFiles/Sphinx.dir/all] Error 2
gmake: *** [Makefile:136: all] Error 2
---
Failed   <<< ddsrouter_docs [0.69s, exited with code 2]
Aborted  <<< fastrtps [3.22s]                            
                                 
Summary: 4 packages finished [4.03s]
  1 package failed: ddsrouter_docs
  1 package aborted: fastrtps
  2 packages had stderr output: ddsrouter_docs ddsrouter_yaml_validator
  5 packages not processed

**Comment by jparisu** [comment]
Hi @iekhanx ,

It seems like you miss some dependencies to compile the sphinx documentation.
Try installing the missing package by: `pip install sphinx-rtd-theme` .
All the requirements to install this documentation are in a file in the respository (https://raw.githubusercontent.com/eProsima/DDS-Router/main/docs/requirements.txt), in case you want to create a python virtual environment.

However, you **do not need to compile the documentation** in order to use the DDS-Router.
This documentation is already available online: https://eprosima-dds-router.readthedocs.io/en/latest/
Compile the project using the following command: `colcon build --packages-select-skip ddsrouter_docs` to skip compiling this documentation package.

Please, keep us updated with your advances and your opinions. We are trying to create a user friendly installation and application, and every feedback is welcomed.

**Comment by iekhanx** [comment]
Hello there,
I am trying to evaluate DDS-Router product in its entirety I manage to build 'GTest' software while placing .in ./install/ directory DDS-Router main directory.
What I am suppose to be TESTING?  Cause this test seem to run fast and which executable file get executed during the TEST?
as my requirement is to test the behaviour of DDS-Router as it route the traffic over the network.
Hence I have two computer at my disposal and third on is on standby I am trying to make sens out of it currently all the activity taking place on the main server but I would like to connect the client workstation and channelling the messages across third computer?
The following statement issued on server host machine and result as shown below:

**$ colcon test --packages-select-regex ddsrouter --event-handler=console_direct+**

Starting >>> ddsrouter_cmake
Starting >>> ddsrouter_yaml_validator
Finished <<< ddsrouter_cmake [0.26s]                                                        
Starting >>> ddsrouter_utils
Starting >>> ddsrouter_docs
Finished <<< ddsrouter_utils [0.03s]
Starting >>> ddsrouter_event                                                               
Finished <<< ddsrouter_docs [0.04s]
Finished <<< ddsrouter_event [0.03s]
Starting >>> ddsrouter_core
Finished <<< ddsrouter_core [0.04s]
Starting >>> ddsrouter_yaml
Finished <<< ddsrouter_yaml [0.04s]                                                        
Starting >>> ddsrouter_tool
Finished <<< ddsrouter_tool [0.04s]
============================= test session starts ==============================
platform linux -- Python 3.10.4, pytest-7.1.1, pluggy-1.0.0
cachedir: /home/imrank/DDS-Router/src/build/ddsrouter_yaml_validator/.pytest_cache
rootdir: /home/imrank/DDS-Router/src/ddsrouter/tools/ddsrouter_yaml_validator
plugins: colcon-core-0.7.1, cov-3.0.0, repeat-0.9.1, rerunfailures-10.2, xonsh-0.11.0
collecting ...                                         
collected 2 items                                                              

tests/ddsrouter_yaml_validator_test.py ..                                [100%]

- generated xml file: /home/imrank/DDS-Router/src/build/ddsrouter_yaml_validator/pytest.xml -
============================== 2 passed in 0.14s ===============================
Finished <<< ddsrouter_yaml_validator [0.82s]          

Summary: 8 packages finished [1.19s]

Your response greatly appriciated, 
Thank you,
IK

**Reply by jparisu** [reply]
It looks like you missed the CMake option `-DBUILD_TESTS=ON` when compiling the project, and so the tests are not installed, thus not executed.
As the package `ddsrouter_yaml_validator` is a python package, it does not require a CMake option and thus the tests are correctly installed and executed.

Try to compile with command `colcon build --cmake-args -DBUILD_TESTS=ON` and run the tests again.
If you want some tips to build and execute the tests, you could take a look to the CI workflow on how the project is installed and the tests run:
https://github.com/eProsima/DDS-Router/blob/main/.github/workflows/test.yml

Nevertheless, we are refactoring and improving the CMake options system (it seems like this is what you are struggling with).
Check this new PR for a new way of managing the CMake optinos along the packages: https://github.com/eProsima/DDS-Router/pull/192

---

About checking the DDS-Router in different devices, there is no automatic test to do so.
You can try executing the examples in the documentation (https://eprosima-dds-router.readthedocs.io/en/latest/rst/examples/change_domain_example.html) and check they work as expected.
